src/network/netlink_netfilter.rs

```diff
@@ -0,0 +1,278 @@
+
+use crate::{
+    error::{ErrorWrap, NetavarkError, NetavarkResult},
+    network::{
+        netlink::{expect_netlink_result, function, NetlinkFamily, Socket},
+        types::PortMapping,
+    },
+};
+use netlink_packet_core::{NLM_F_ACK, NLM_F_DUMP};
+use netlink_packet_netfilter::{
+    conntrack::ConntrackMessage, NetfilterHeader, NetfilterMessage, ProtoFamily,
+};
+use netlink_sys::protocols::NETLINK_NETFILTER;
+
+pub struct NetlinkNetfilter;
+
+impl NetlinkFamily for NetlinkNetfilter {
+    const PROTOCOL: isize = NETLINK_NETFILTER;
+    type Message = NetfilterMessage;
+}
+
+// Represents the 5-tuple for a single direction of a connection flow.
+#[derive(Clone)]
+pub struct FlowTuple {
+    pub src_ip: IpAddr,
+    pub dst_ip: IpAddr,
+    pub src_port: u16,
+    pub dst_port: u16,
+    pub protocol: Protocol,
+}
+impl FlowTuple {
+    pub fn is_ipv6(&self) -> bool {
+        self.src_ip.is_ipv6() && self.dst_ip.is_ipv6()
+    }
+}
+
+// Represents a conntrack entry, detailing the original and reply flows.
+#[derive(Clone)]
+pub struct ConntrackFlow {
+    pub origin: Option<FlowTuple>,
+    pub reply: Option<FlowTuple>,
+}
+
+// Converts a `ConntrackFlow` into a vector of `ConntrackNla`
+// attributes suitable for a netlink message payload.
+impl ConntrackFlow {
+    fn to_attributes(self: ConntrackFlow) -> Vec<ConntrackNla> {
+        fn to_tuple_attributes(flow_tuple: &FlowTuple) -> Vec<Tuple> {
+            let ip_attributes = vec![
+                IPTuple::SourceAddress(flow_tuple.src_ip),
+                IPTuple::DestinationAddress(flow_tuple.dst_ip),
+            ];
+
+            let proto_attributes = vec![
+                ProtoTuple::Protocol(flow_tuple.protocol),
+                ProtoTuple::SourcePort(flow_tuple.src_port),
+                ProtoTuple::DestinationPort(flow_tuple.dst_port),
+            ];
+            vec![Tuple::Ip(ip_attributes), Tuple::Proto(proto_attributes)]
+        }
+
+        let mut attributes = Vec::new();
+
+        if let Some(ref origin_flow) = self.origin {
+            let origin_attributes = to_tuple_attributes(origin_flow);
+            if !origin_attributes.is_empty() {
+                attributes.push(ConntrackNla::CtaTupleOrig(origin_attributes));
+            }
+        }
+
+        if let Some(ref reply_flow) = self.reply {
+            let reply_attributes = to_tuple_attributes(reply_flow);
+            if !reply_attributes.is_empty() {
+                attributes.push(ConntrackNla::CtaTupleReply(reply_attributes));
+            }
+        }
+
+        attributes
+    }
+    fn is_ipv6(&self) -> NetavarkResult<bool> {
+        match (&self.origin, &self.reply) {
+            (Some(origin), Some(reply)) => {
+                let origin_is_v6 = origin.is_ipv6();
+
+                if origin_is_v6 != reply.is_ipv6() {
+                    Err(NetavarkError::Message(
+                        "The families are different".to_string(),
```

Comment:
Nit: if you need to repush, can you change this to "Mismatched IP families in conntrack flow" to make it more clear?